bom/generate_warg_bom.py

```diff
@@ -1,49 +1,101 @@
+
+from xml.etree import ElementTree as ET
+
+import sys, os
+import csv
+
+def main():
+    if(len(sys.argv) != 3):
+        print "Make sure there are only two arguments, the netlist xml & the output name"
+    filepath, netlist_xml, output_filename = sys.argv
+    
+    #found_parts is for seeing if you already added that part
+    #output_dict is the dictionary that gets written to the csv file
+    found_parts = []
+    output_dict = []
+    keylist = ['ref', 'quantity']
+
+    #Parse the netlist and store it in an lxml tree
+    tree = ET.parse(netlist_xml)
+
+    #Get a list of all the keys for the CSV at the end
+    for part_element in tree.findall(".//*[@name='part_num']...."):
+        print part_element
+        if part_element.tag == 'comp':
+            fields = part_element.find('fields')
+            for child in fields:
+                if child.attrib['name'] not in keylist:
+                    keylist.append(child.attrib['name'])
+
+
+    #Traverse over every element with a WARG Part Number 
+    for part_element in tree.findall(".//*[@name='part_num']...."):
+        #Check if the element is under a component or a library element
+        if part_element.tag == 'comp':
+            part = dict()
+            fields = part_element.find('fields')
+            for child in fields.findall('field'):
+                part.update({child.attrib['name']:child.text})
+
+            ref = part_element.attrib['ref']
+            
+            #Find a matching part to add a quantity
+            matching_part = {}
+            for check in found_parts:
+                #Check if they both have the same keys
+                if check.keys() == part.keys():
+                    #Check if all the values are the same
+                    for key in check.keys():
+                        if check[key] != part[key]:
+                            break
+                        matching_part = check
+
+            if matching_part == {}:      
+                found_parts.append(part.copy())
+                part.update({'quantity':1, 'ref':ref})
+                output_dict.append(part)
+            else:
+                output_dict[found_parts.index(part)]['quantity'] += 1
+                output_dict[found_parts.index(part)]['ref'] += (', ' + ref)
+    reader = csv.DictReader(open('component_library.csv', 'rb'))
+
+    #Add the component library keys to the keylist
+    for key in reader.fieldnames:
+        if key not in keylist:
+            keylist.append(key)
+
+    #Add the values from the component_library to the output dictionary
+    for line in reader:
+        if "part_num" not in line:
+            print "line doesn't have part_num! line: ", line
+        for part in output_dict:
+            if part['part_num'] == line['part_num']:
+                part.update(line)
+
+    #Check for parts missing from the library
+    missing_parts = []
+    for part in output_dict:
+        try:   
+            if part['Distributer #'] != '':

```

Comment:
Parts can be in the WARG library and not have distributer part numbers. Maybe add some logic to line 74 to check if there is a matching part number in the component library.